### DISCUSSION TITLE
How do I create an admin role?

### DISCUSSION THREAD

**Discussion by engineervix** [hagopj13/node-express-boilerplate#60] [Q&A] [answered]
I can register a _user_ role like this

```
POST http://localhost:3001/v1/auth/register
content-type: application/json

{
    "name": "John Doe",
    "email": "john.doe@example.com",
    "password": "Password1"
}
```

How can I register an _admin_ role?

**Comment by hagopj13** [comment, marked as answer]
The `register` endpoint can be used only to register as a user.
To create an admin, you should use the `POST /v1/users` (which is the create user endpoint) and specify `role: 'admin'` in the body. You need to be logged in as an admin to be allowed to do that. This means that the first admin should be created manually in the db.

**Reply by engineervix** [reply]
I see! Makes sense! Thanks a lot.

**Reply by ymadhubabu** [reply]
I could not get to it though I tried the ways u mentioned..kindly elaborate
Thanks in advance

**Reply by ymadhubabu** [reply]
> I see! Makes sense! Thanks a lot.

Hi ...Could u do that ..if so, please let me know his how u done it

**Reply by ymadhubabu** [reply]
i tried it this way as shown in snaps....but stuck with authentication errors when i tried to create new admin user through post v1/users and also i was not able to register a user when i connected with db admin credentials....

But i could register as user when i connected without db admin credentials


please help me

![image](https://user-images.githubusercontent.com/22412676/123193227-ffc7e880-d4c1-11eb-8d4d-e844b462cdac.png)
![image](https://user-images.githubusercontent.com/22412676/123193303-2554f200-d4c2-11eb-87a8-f8a426687a6e.png)
![image](https://user-images.githubusercontent.com/22412676/123193367-4ddcec00-d4c2-11eb-800a-f6d72d2d8e31.png)
![image](https://user-images.githubusercontent.com/22412676/123193400-5df4cb80-d4c2-11eb-97ee-a9b2ff6d5578.png)
![image](https://user-images.githubusercontent.com/22412676/123193429-6baa5100-d4c2-11eb-8a3c-bdb73a785b91.png)
![image](https://user-images.githubusercontent.com/22412676/123193509-8c72a680-d4c2-11eb-99ab-6bf82c8f21ff.png)

**Reply by ymadhubabu** [reply]
this is the error messsage shown on the mongod server console  when i connected as db admin 


_**{"t":{"$date":"2021-06-24T08:09:57.093+05:30"},"s":"I",  "c":"ACCESS",   "id":20250,   "ctx":"conn16","msg":"Authentication succeeded","attr":{"mechanism":"SCRAM-SHA-256","speculative":true,"principalName":"mongoadmin","authenticationDatabase":"admin","remote":"127.0.0.1:65029","extraInfo":{}}}
{"t":{"$date":"2021-06-24T08:09:57.114+05:30"},"s":"I",  "c":"ACCESS",   "id":20436,   "ctx":"conn16","msg":"Checking authorization failed","attr":{"error":{"code":13,"codeName":"Unauthorized","errmsg":"not authorized on admin to execute command { getCmdLineOpts: 1, lsid: { id: UUID(\"ef7867a2-3c49-4832-b1b0-b91af3faeff6\") }, $db: \"admin\" }"}}}**_

**Reply by hagopj13** [reply]
@ymadhubabu a much easier way would be to edit the code and remove the `auth` middleware from the create user endpoint. Then simply create a user with the role `admin` (as described above) and there you have your admin

**Reply by ymadhubabu** [reply]
Thank you so much

**Reply by sakydogalo** [reply]
I don't know how to manage users. I want to list users and or add  users. I previously registered an user with the user role. through mongoDB Compas, I have modified the role of that user as admin. I log in as admin user and receive a token:

I log in as admin user with:
```
curl -X 'POST' \
  'http://localhost:3000/v1/auth/login' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "email": "fake@example.com",
  "password": "password1"
}'
```
and receive the token:
```
200 | Response body
{
  "user": {
    "role": "admin",
    "isEmailVerified": true,
    "name": "fake name",
    "email": "fake@example.com",
    "id": "6351bbe44914b069e246e26b"
  },
  "tokens": {
    "access": {
      "token": "...",
      "expires": "2022-10-20T22:05:17.325Z"
    },
    "refresh": {
      "token":   "<<<< this token >>>>",
      "expires": "2022-11-19T22:35:17.325Z"
    }
  }
}
```
I enter the token in the authorization and get this instruction to list users:
```
curl -X 'GET' \
  'http://localhost:3000/v1/users?limit=10&page=1' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer <<<< this token >>>>'
```
But the result is bad 
```
error: Error: Please authenticate
    at /home/saky/Dev/Lab04.D3viL/00-03-partes/v001/src/middlewares/auth.js:15:19
    at JwtStrategy.strategy.error (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport/lib/middleware/authenticate.js:358:18)
    at verified (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport-jwt/lib/strategy.js:111:41)
    at JwtStrategy.jwtVerify [as _verify] (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/src/config/passport.js:22:5)
    at /home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport-jwt/lib/strategy.js:123:34
    at /home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/jsonwebtoken/verify.js:223:12
    at getSecret (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/jsonwebtoken/verify.js:90:14)
    at Object.module.exports [as verify] (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/jsonwebtoken/verify.js:94:10)
    at Function.module.exports [as JwtVerifier] (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport-jwt/lib/verify_jwt.js:4:16)
    at /home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport-jwt/lib/strategy.js:104:25
    at JwtStrategy._secretOrKeyProvider (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport-jwt/lib/strategy.js:40:13)
    at JwtStrategy.authenticate (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport-jwt/lib/strategy.js:99:10)
    at attempt (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport/lib/middleware/authenticate.js:366:16)
    at authenticate (/home/saky/Dev/Lab04.D3viL/00-03-partes/v001/node_modules/passport/lib/middleware/authenticate.js:367:7)
    at /home/saky/Dev/Lab04.D3viL/00-03-partes/v001/src/middlewares/auth.js:36:109
    at new Promise (<anonymous>)
error: GET /v1/users?limit=10&page=1 401 - 5.473 ms - message: Please authenticate
```
Where am I going wrong?

**Reply by JOASJSF** [reply]
Same situation, were you able to solve it?

**Reply by Rushikatkar** [reply]
I am also facing same issue, if anybody knows about the solution of this problem then please help..
